client/components/external-gridicons/index.js

```diff
@@ -0,0 +1,65 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import svg4everybody from 'svg4everybody';
+
+/**
+ * Internal dependencies
+ */
+import { iconsThatNeedOffset, iconsThatNeedOffsetX, iconsThatNeedOffsetY } from './icons-offset';
+
+function needsOffset( name, icons ) {
+	return icons.indexOf( name ) >= 0;
+}
+
+const isBrowser = typeof window !== 'undefined';
+if ( isBrowser ) {
+	// Polyfill SVG external content support. Noop in the evergreen build.
+	svg4everybody();
+}
+
+function ExternalGridicons( props ) {
+	const { size = 24, icon, onClick, className, ...otherProps } = props;
+	const isModulo18 = size % 18 === 0;
+
+	// Using a missing icon doesn't produce any errors, just a blank icon, which is the exact intended behaviour.
+	// This means we don't need to perform any checks on the icon name.
+	const iconName = `gridicons-${ icon }`;
+	const offsetClasses = isModulo18
+		? [
+				needsOffset( iconName, iconsThatNeedOffset ) ? 'needs-offset' : false,
+				needsOffset( iconName, iconsThatNeedOffsetX ) ? 'needs-offset-x' : false,
+				needsOffset( iconName, iconsThatNeedOffsetY ) ? 'needs-offset-y' : false,
+		  ]
+		: [];
+	const iconClass = [ 'gridicon', iconName, className, ...offsetClasses ]
+		.filter( Boolean ) // Remove all falsy values.
+		.join( ' ' );
+
+	return (
+		<svg
+			xmlns="http://www.w3.org/2000/svg"
+			viewBox="0 0 24 24"
+			className={ iconClass }
+			height={ size }
+			width={ size }
+			onClick={ onClick }
+			{ ...otherProps }
+		>
+			<use xlinkHref={ `/calypso/images/gridicons/sprite.svg#${ iconName }` } />
```

Comment:
do we need to cachebust this somehow?

Reply:
maybe use `file-loader` to get the url, which can bake the hash in?

Reply:
Yes, we'll need to have both a mechanism for keeping gridicons up-to-date from the gridicons repo, and a mechanism to keep users' copies up-to-date (with some form of cache-busting, yes).

I intentionally omitted any attempt at either of those in this PR, since I'm just looking for validation on the general approach, for now.

Reply:
I think the approach is good. How do you want to proceed with this?

Reply:
> I think the approach is good. How do you want to proceed with this?

Here are my initial thoughts:

Work in gridicons repo:
- Modify build to include spritemap SVG as part of the gridicons NPM package.
- Modify build to include offset configuration as part of the gridicons NPM package (see the `icons-offset.js` file in this PR)
- For both of the above changes, we could keep the entry points as they are, so that existing consumers aren't affected. These would be extra files on the same package that would have to be required explicitly.
- Deploy new NPM package.

Work in Calypso repo:
- Increment the gridicons NPM package version to match the new release.
- Include the copying of the spritemap as part of the build process. Cache-bust spritemap based on the NPM package version, if that's somehow possible.
- Implement the rest of the approach as per this PR.

How does that sound?

Reply:
Sounds good. I think we have an earlier candidate for svg4everybody over in https://github.com/Automattic/wp-calypso/pull/32236. @griffbrad Do you need to coordinate that work at all?

Reply:
@griffbrad 's approach for the Material icons looks quite different, in that he's wrapping each icon in a React component, created at build-time through SVGR. That's the approach previously used in gridicons (SVG-in-JS), albeit with a simplified pipeline. It's a good approach if you're using an icon or two, but if you're using a large number of them (like we do in gridicons), the cost of wrapping SVG in JS starts taking a toll.

`svg4everybody` isn't needed in @griffbrad 's case, since as far as I can tell he's not using SVG `<use>` pointing to external resources, nor actually loading anything from SVG files at runtime.

Reply:
> svg4everybody isn't needed in @griffbrad 's case, since as far as I can tell he's not using SVG <use> pointing to external resources, nor actually loading anything from SVG files at runtime.

Right, I'd like to consolidate down to one method of loading external SVGs rather than both of these and I'd rather avoid transpiling svg to react.

Reply:
Gotcha. Feel free to add me as a reviewer in any SVG-loading PRs, then, and please feel free to reuse any of the code in this PR! I'll see if I can get to work on the gridicons side of things before continuing here in Calypso.